improve quantized error checking for structured kernels

bdhirsh opened the pull request:
This PR improves the error messaging for structured operators that don't have a quantized backend implementation.

`at::native_batch_norm` does not currently have a quantized implementation. The current error when you try to run that op with quantized tensors is:
```
"Could not run 'aten::native_batch_norm' with arguments from the 'QuantizedCPU' backend"
```

Making the operator structured (in https://github.com/pytorch/pytorch/pull/67343) makes the error message worse though:
```
Could not run 'aten::empty.memory_format' with arguments from the 'QuantizedCPU' backend
```

Why does that happen? After making `native_batch_norm` structured, it gets a code-generated `CompositeExplicitAutograd` kernel. The codegen'd kernel just constructs an empty tensor using `at::empty()` and calls into the `out=` kernel.

This is useful for most backends because they can choose to implement only the `out=` operator if they want to, and get a composite out-of-place kernel that would work automatically.

This ends up being bad for quantization though because quantization doesn't support `at::empty()`, so we end up breaking with a (pretty useless) error message.

In this PR, I add a dynamic check for if the operator we're trying to call is meant to be quantized, and if it is I return an undefined tensor. This lets us try to dispatch all the way to `native_batch_norm.out`, at which point the dispatcher complains with a (friendlier) error message.

**Testing**

I tested the PR by rebasing https://github.com/pytorch/pytorch/pull/67343 on top of it and running `python test/test_quantization.py TestFuseEager.test_fuse_module_train` to confirm that the error message now mentions `native_batch_norm` instead of `empty.memory_format`

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#71928 improve quantized error checking for structured kernels**

Differential Revision: [D33823417](https://our.internmc.facebook.com/intern/diff/D33823417)

bdhirsh committed:
improve quantized error checking for structured kernels

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33823417).

wconstab commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -56,14 +56,29 @@ def gen_create_out_helper(backend_index: BackendIndex) -> List[str]:
+    // If the out-of-place op doesn't have a quantized kernel (which is why we ended up here),
+    // then the out= variant of the op typically won't either, so we'll error out with
+    // a nicer error message, complaining that the out= operator doesn't have a quantized implementation.
+    return at::Tensor();
```
so by doing this, does it guarantee that composite out= operator (e.g. structured native_batch_norm) fails, just by virtue of noticing one of its inputs is undefined?  

i'm not following how this results in an error message that implicates missing an implementation for quantized backend.  it seems more like it would just fail due to an undefined input tensor.

but I have a feeling you're one step ahead of me and this is probably fine :)

bdhirsh replied:
It actually doesn't guarantee that the composite out= operator fails.

I'm basically assuming that if `{op}` doesn't have a quantized kernel, then `{op}.out` won't have one either. If you call `{op}` with quantized tensors, then because of the codegen'd kernel here, we'll funnel you to the `{op}.out` operator (and pass it an undefined output tensor). As long as `{op}.out` doesn't have a quantized implementation, then we'll fail in the *dispatcher* by complaining that `{op}.out` doesn't have a quantized implementation.

Technically, if there were an op that the quantization team happened to write a kernel for the out= variant but not the out-of-place variant, then with this PR it would have to handle undefined tensors properly if you wanted to be able to call the out-of-place variant. That's strictly better than today though, where calling the out-of-place variant will just unconditionally fail (because we try to call `at::empty()`, which will never work for quantized).

Long term, a better fix would be to add real support for quantized backends to structured kernels (there's actually an open issue for it: https://github.com/pytorch/pytorch/issues/52680). That is definitely more involved though, so this PR will just avoid making things worse until then.

Let me know if

wconstab replied:
ahh ok that makes sense.  ok, then i agree its strictly better. thanks!

bdhirsh committed:
Update on "improve quantized error checking for structured kernels"

This PR improves the error messaging for structured operators that don't have a quantized backend implementation.

`at::native_batch_norm` does not currently have a quantized implementation. The current error when you try to run that op with quantized tensors is:
```
"Could not run 'aten::native_batch_norm' with arguments from the 'QuantizedCPU' backend"
```

Making the operator structured (in https://github.com/pytorch/pytorch/pull/67343) makes the error message worse though:
```
Could not run 'aten::empty.memory_format' with arguments from the 'QuantizedCPU' backend
```

Why does that happen? After making `native_batch_norm` structured, it gets a code-generated `CompositeExplicitAutograd` kernel. The codegen'd kernel just constructs an empty tensor using `at::empty()` and calls into the `out=` kernel.

This is useful for most backends because they can choose to implement only the `out=` operator if they want to, and get a composite out-of-place kernel that would work automatically.

This ends up being bad for quantization though because quantization doesn't support `at::empty()`, so we end up breaking with a (pretty useless) error message.

In this PR, I add a dynamic check for if the operator we're trying to call is meant to be quantized, and if it is I return an undefined tensor. This lets us try to dispatch all the way to `native_batch_norm.out`, at which point the dispatcher complains with a (friendlier) error message.

**Testing**

I tested the PR by rebasing https://github.com/pytorch/pytorch/pull/67343 on top of it and running `python test/test_quantization.py TestFuseEager.test_fuse_module_train` to confirm that the error message now mentions `native_batch_norm` instead of `empty.memory_format`


Differential Revision: [D33823417](https://our.internmc.facebook.com/intern/diff/D33823417)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33823417).

bdhirsh commented:
@bhosmer changed the comment a bit and also defined a more generic helper function, so the codegen doesn't need to directly add an "if quantized" check.
